storage: TestReopenConnection failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/62c976dafc7f48f0a52142310a8fdc42908c30a5

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestReopenConnection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=961079&tab=buildLog

```
=== RUN   TestReopenConnection
I181012 05:39:46.584364 109396 rpc/nodedialer/nodedialer.go:91  unable to connect to n2: initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:43009: connect: connection refused"
I181012 05:39:46.584734 109396 rpc/nodedialer/nodedialer.go:91  unable to connect to n3: initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:43009: connect: connection refused"
I181012 05:39:46.584791 109292 gossip/gossip.go:844  [n1] removing n2 which was at same address ({tcp 127.0.0.1:43009}) as new node {3 {tcp 127.0.0.1:43009}   0.0  0 []}
I181012 05:39:46.584820 109292 gossip/gossip.go:806  [n1] removed n2 from gossip
W181012 05:39:47.674552 109433 storage/raft_transport.go:578  creating batch client for node 3 failed: rpc error: code = Canceled desc = grpc: the client connection is closing
--- FAIL: TestReopenConnection (1.10s)
	raft_transport_test.go:582: replacement node still unhealthy

```

--- a/pkg/rpc/nodedialer/nodedialer.go
+++ b/pkg/rpc/nodedialer/nodedialer.go
@@ -103,6 +103,12 @@ func (n *Dialer) Dial(ctx context.Context, nodeID roachpb.NodeID) (_ *grpc.Clien
 		return nil, err
 	}
 	breaker.Success()
+	// This really shouldn't be necessary, but Success() isn't guaranteed to
+	// close the breaker and the upstream vendored repo is dead, so just work
+	// around it by resetting the breaker manually if it's still tripped.
+	if breaker.Tripped() {
+		breaker.Reset()
+	}
 	return conn, nil
 }
 

